
# pseudobulk, presence, clr, and seg norm data

```{r}
library(dplyr)
library(tidyr)
library(reshape2)
library(Matrix)
library(writexl)
```

```{r include=FALSE}
file.copy('../../preprocess/generate_single_cell_data/lm_glycogene_pseudopresence.Rds', 'lm_glycogene_pseudopresence.Rds')
pseudopresence_model.lm <- readRDS('lm_glycogene_pseudopresence.Rds')
file.remove('lm_glycogene_pseudopresence.Rds')

compute_sample_pop <- function(pseudobulk, mu_model.lm, sd_model.lm) {

  mu <-
    stats::predict(
      mu_model.lm,
      newdata = data.frame(pseudobulk_mean = pseudobulk),
      se.fit = TRUE,
      level = 0.95,
      interval = "confidence")$fit[,'fit']

  s <-
    stats::predict(
      sd_model.lm,
      newdata = data.frame(pseudobulk = pseudobulk),
      se.fit = TRUE,
      level = 0.95,
      interval = "confidence")$fit[,'fit']

  return(list(mu = mu, s = s))

}
```

```{r include = FALSE}
## computing pvalue for each glycogene on each cluster
compute_pseudopresence_pvalue <- function(pseudobulk, mu, s, n) {

  # n chosen based on the number of pseudobulks that model is fit to
  ts  <- (mu - pseudobulk)/(s/sqrt(n))
  p   <- 2*pt(-abs(ts), n-1)

  return(p)

}
```

```{r include=FALSE}

file.copy('../../preprocess/generate_single_cell_data/lm_chromium10x_mupb_norm_nziqr_threshold_cutoff.Rds', 'lm_chromium10x_mupb_norm_nziqr_threshold_cutoff.Rds')
expression_model.lm <- readRDS('lm_chromium10x_mupb_norm_nziqr_threshold_cutoff.Rds')
cutoff.int <- mean(expression_model.lm$xy.df$y[1:7])
file.remove('lm_chromium10x_mupb_norm_nziqr_threshold_cutoff.Rds')

compute_threshold <- function(ncells) {

  s <-
    stats::predict(
      expression_model.lm$model2.lm,
      newdata = data.frame(x = ncells),
      se.fit = TRUE,
      level = 0.95,
      interval = "confidence")$fit[,'fit']

  return(s)

}
```

```{r include=FALSE}
compute_expression_pvalue <- function(pseudobulk, threshold) {

  n   <- 7 # chosen based on the number of pseudobulk sizes that model is fit to
  ts  <- (cutoff.int - pseudobulk)/(threshold/sqrt(n))
  p   <- 2*pt(-abs(ts), n-1)

  return(p)

}
```

# intersect function

```{r}

compute_intersects <- function(test_data, reference_data) {

  rownames(reference_data) <- reference_data$gene
  reference_data$gene <- NULL

  rownames(test_data) <- test_data$gene
  test_data$gene <- NULL

  reference_data <- as.matrix(reference_data)

  ## ensuring genes of reference_data and test_data are in order ----
  ordering <-
    sort(unique(c(
      rownames(test_data),
      rownames(reference_data))))

  test_data <- test_data[ordering,]
  reference_data <- reference_data[ordering,]

  ## computing value for specified quantile cutoffs on reference_data
  expression_quantiles <- apply(
    reference_data, 1,
    function(x) {
      # type 1 equivalent to inverse of ecdf function
      quantile(x, probs = seq(0,1,0.1), type = 1, na.rm = T)
    })

  ## computing predicted expression for each gene at each cutoff ----
  predicted_expression <-
      mapply(
        function(sample_library, cutoffs) {
          ecdf(cutoffs)(sample_library)
        },
        data.frame(t(test_data)),
        data.frame(t(reference_data)))

  rownames(predicted_expression) <- colnames(test_data)
  predicted_expression <- reshape2:::melt.matrix(
    predicted_expression,
    varnames = c('subset','gene'),
    value.name = 'percentile')

  predicted_expression

  list(predicted_expression = predicted_expression)

}

```

# dynamic range function

```{r}

compute_ranges <- function(reference_data) {

  rownames(reference_data) <- reference_data$gene
  reference_data$gene <- NULL

  reference_data <- as.matrix(reference_data)

  ## ensuring genes of reference_data and test_data are in order ----
  ordering <- sort(unique(rownames(reference_data)))

  reference_data <- reference_data[ordering,]

  ## computing value for specified quantile cutoffs on reference_data
  expression_quantiles <- apply(
    reference_data, 1,
    function(x) {
      # type 1 equivalent to inverse of ecdf function
      quantile(x, probs = seq(0,1,0.1), type = 1, na.rm = T)
    })

  list(expression_quantiles = expression_quantiles)

}

```

# loading glycogenes

```{r}
file.copy('../../preprocess/generate_single_cell_data/seg_lookup.Rds', 'seg_lookup.Rds')
segs = unique(names(readRDS('seg_lookup.Rds')))
file.remove('seg_lookup.Rds')

file.copy('../../preprocess/generate_single_cell_data/mouse_orthology_human_genes.tsv', 'mouse_orthology_human_genes.tsv')
ortho_trans_table=with(read.delim('mouse_orthology_human_genes.tsv'),setNames(mouse_symbol,human_symbol))
file.remove('mouse_orthology_human_genes.tsv')

file.copy('../../preprocess/generate_single_cell_data/glycogenes.tsv', 'glycogenes.tsv')
glycogenes=read.delim('glycogenes.tsv',header=F)$V1
file.remove('glycogenes.tsv')
```

# loading bulk GTEx data

```{r}

file.copy('../../preprocess/generate_bulk_data/tissue_results_log.Rds', 'tissue_results_log.Rds')
GTEx_bulk <- readRDS('tissue_results_log.Rds')
GTEx_bulk <- subset(GTEx_bulk, ref_gene == 'all' & samples > 10 & log_cpm > log(2))
GTEx_bulk$ref_gene <- NULL
rownames(GTEx_bulk) <- NULL
file.remove('tissue_results_log.Rds')

```

# loading bulk TCGA data

```{r}

file.copy('../../preprocess/generate_bulk_data/tissue_results_log_tcga.Rds', 'tissue_results_log_tcga.Rds')
TCGA_bulk <- readRDS('tissue_results_log_tcga.Rds')
TCGA_bulk <- subset(TCGA_bulk, ref_gene == 'all' & samples > 10 & log_cpm > log(2))
TCGA_bulk$ref_gene <- NULL
rownames(TCGA_bulk) <- NULL
file.remove('tissue_results_log_tcga.Rds')

```

# joining bulk datasets

```{r}

bulk <- cbind.data.frame(GTEx_bulk, source = 'GTEx')
bulk <- subset(
  rbind.data.frame(
    cbind.data.frame(GTEx_bulk, source = 'GTEx'),
    cbind.data.frame(TCGA_bulk, source = 'TCGA')),
    !(gene %in% segs))

```

```{r}

bulk_glycogenes <- sort(unique(bulk$gene))

```

```{r}

# remove(GTEx_bulk)
# remove(TCGA_bulk)

```

# Tabula Sapiens (human)

## loading seg normalised data (glycogenes)

```{r}

file.copy('../../preprocess/generate_single_cell_data/pseudobulk_log_seg_norm.Rds', 'pseudobulk_log_seg_norm.Rds')
glycogene_segnorm.df <- readRDS("pseudobulk_log_seg_norm.Rds")
file.remove('pseudobulk_log_seg_norm.Rds')

```

```{r}

glycogene_human_segnorm.df <- glycogene_segnorm.df[with(glycogene_segnorm.df, source == 'human'),]

```

```{r}

remove(glycogene_segnorm.df)

```

## loading clrs (all genes)

```{r}

file.copy('../../preprocess/generate_single_cell_data/alldata_clean_human_clr_sparse.Rds', 'alldata_clean_human_clr_sparse.Rds')
alldata_clean_human_clr_sparse.sm <- readRDS('alldata_clean_human_clr_sparse.Rds')
file.remove('alldata_clean_human_clr_sparse.Rds')

```

### filtering for glycogenes only

```{r}

glycogene_clean_human_clr_sparse.sm <- alldata_clean_human_clr_sparse.sm[,glycogenes]

```

```{r}

remove(alldata_clean_human_clr_sparse.sm)

```

### converting to data.frame

```{r}

glycogene_human_clr.df <-
  setNames(
    subset(reshape2:::melt.matrix(glycogene_clean_human_clr_sparse.sm, as.is = T), value != 0),
    c('tissue','gene','clr'))

```

```{r}

remove(glycogene_clean_human_clr_sparse.sm)

```

### relabeling clusters to match seg norm data

```{r}

glycogene_human_clr.df$tissue <- gsub('_cluster.*$', '', glycogene_human_clr.df$tissue)

```

## loading pseudobulpresence data (all genes)

```{r}

file.copy('../../preprocess/generate_single_cell_data/alldata_pseudobulkpresence.Rds', 'alldata_pseudobulkpresence.Rds')
alldata_pseudobulkpresence.df <- readRDS('alldata_pseudobulkpresence.Rds')
file.remove('alldata_pseudobulkpresence.Rds')

```

```{r}

cutoff.int <- attributes(alldata_pseudobulkpresence.df)$cutoff.int

```

### filtering for glycogenes only

```{r}

glycogene_human_pseudobulkpresence.df <- alldata_pseudobulkpresence.df[with(alldata_pseudobulkpresence.df, gene %in% glycogenes),]

```

```{r}

remove(alldata_pseudobulkpresence.df)

```

### relabeling clusters to match seg norm data

```{r}

glycogene_human_pseudobulkpresence.df <- dplyr::rename(glycogene_human_pseudobulkpresence.df, tissue=cluster, samples=cellnum)
glycogene_human_pseudobulkpresence.df$tissue <- gsub('_cluster.*$', '', glycogene_human_pseudobulkpresence.df$tissue)

```

## merging datasets

```{r}

glycogene_human.df <- Reduce(
  merge,
  list(
    glycogene_human_pseudobulkpresence.df,
    glycogene_human_clr.df,
    glycogene_human_segnorm.df))

```

```{r}

remove(glycogene_human_pseudobulkpresence.df)
remove(glycogene_human_clr.df)
remove(glycogene_human_segnorm.df)

```

### dropping columns

```{r}

glycogene_human.df$threshold <- NULL
glycogene_human.df$orig.NA <- NULL
glycogene_human.df$source <- NULL

```

### splitting columns

```{r}

glycogene_human.df <- tidyr::separate(glycogene_human.df, 'tissue', into = c('tissue', 'cell'), sep = ':')

```

### joining columns

```{r}

glycogene_human.df$pseudopresence.pred_mu <- with(
  glycogene_human.df,
  round(pseudopresence.pred_mu, 2))

glycogene_human.df$pseudopresence.pred_sd <- with(
  glycogene_human.df,
  round(pseudopresence.pred_sd, 2))

glycogene_human.df <- tidyr::unite(
  glycogene_human.df,
  'pseudopresence.expected',
  pseudopresence.pred_mu,
  pseudopresence.pred_sd,
  sep = '+/-')

```

### adding columns

```{r}

glycogene_human.df$organism <- 'human'

```

```{r}

glycogene_human.df$gene_ortholog <- glycogene_human.df$gene

```

```{r}

glycogene_human.df$pseudopresence.classification <- with(
  glycogene_human.df,
  ifelse(pseudopresence.pval > 1E-5, 'stable','unstable'))

```

```{r}

glycogene_human.df$above_cutoff <- glycogene_human.df$pseudobulk > cutoff.int

```

```{r}

test_data <- glycogene_human.df[
  glycogene_human.df$gene %in% intersect(
    unique(glycogene_human.df$gene),
    unique(glycogene_human.df$gene)),]

reference_data <- test_data

glycogene_human.df <- merge(
  glycogene_human.df,
  compute_intersects(
    test_data = reshape2:::dcast(
      test_data,
      gene ~ paste0(tissue,':',cell),
      value.var = 'rel_diff'),
    reference_data = reshape2:::dcast(
      reference_data,
      gene ~ paste0(tissue,':',cell),
      value.var = 'rel_diff'))$predicted_expression %>%
    tidyr::separate('subset', into = c('tissue', 'cell'), sep = ':') %>%
    dplyr::rename(sc_percentile=percentile),
  by = c('cell','tissue','gene'),
  all.x = T)

```

```{r}

glycogene_human.df$sc_classification <- with(
  glycogene_human.df, cut.default(
    sc_percentile,
    breaks = seq(-0.1,1.1,0.1),
    labels = c(paste0('<',seq(0,100,10),'%'),'>100%')))

```

```{r}

test_data <- glycogene_human.df[
  glycogene_human.df$gene %in% intersect(
    unique(glycogene_human.df$gene),
    unique(bulk$gene[bulk$source == 'GTEx'])),]

reference_data <- bulk[
  bulk$source == 'GTEx' &
  bulk$gene %in% intersect(
    unique(bulk$gene[bulk$source == 'GTEx']),
    unique(glycogene_human.df$gene)),]

glycogene_human.df <- merge(
  glycogene_human.df,
  compute_intersects(
    test_data = reshape2:::dcast(
      test_data,
      gene ~ paste0(tissue,':',cell),
      value.var = 'rel_diff'),
    reference_data = reshape2:::dcast(
      reference_data,
      gene ~ paste0(sample_id,':',tissue,':',source),
      value.var = 'rel_diff'))$predicted_expression %>%
    tidyr::separate('subset', into = c('tissue', 'cell'), sep = ':') %>%
    dplyr::rename(bulk_percentile=percentile),
  by = c('cell','tissue','gene'),
  all.x = T)

```

```{r}

glycogene_human.df$bulk_classification <- with(
  glycogene_human.df, cut.default(
    bulk_percentile,
    breaks = seq(-0.1,1.1,0.1),
    labels = c(paste0('<',seq(0,100,10),'%'),'>100%')))

```

### repeating for cell specific aggregate

#### subset the dataset

```{r}

glycogene_human_cell.df <-
  glycogene_human.df %>% group_by(gene, cell) %>% summarize(
    organism = unique(organism),
    gene_ortholog = unique(gene_ortholog),
    clr = mean(clr),
    pseudobulk = mean(pseudobulk),
    samples = sum(samples),
    pseudopresence = sum(pseudopresence),
    rel_diff = mean(rel_diff))

```

#### apply the cutoff

```{r}

glycogene_human_cell.df$above_cutoff <- glycogene_human_cell.df$pseudobulk > cutoff.int

```

#### compute the threshold

```{r}
glycogene_human_cell.df$threshold <- compute_threshold(glycogene_human_cell.df$samples)
```

#### compute the expression pvalue

```{r}
glycogene_human_cell.df$pvalue <- compute_expression_pvalue(glycogene_human_cell.df$pseudobulk, glycogene_human_cell.df$threshold)
glycogene_human_cell.df$threshold <- NULL
```

#### compute the expression classification

```{r}
glycogene_human_cell.df$`status5E-03` <- with(
  glycogene_human_cell.df,
  addNA(factor(
    ifelse(
      pvalue > 5E-03, 'unpredictable',
      ifelse(pseudobulk > cutoff.int, 'expressed', 'unexpressed')),
    levels = rev(c('unexpressed','unpredictable','expressed')))))
```

#### computing expected pseudopresence mu and sd

```{r}
predicted_percent_pseudopresence <-
  compute_sample_pop(
    glycogene_human_cell.df$pseudobulk,
    pseudopresence_model.lm$glycogene_mean_fit.lm,
    pseudopresence_model.lm$glycogene_sd_fit.lm)

```

```{r}

glycogene_human_cell.df$pseudopresence.pred_mu <- (predicted_percent_pseudopresence$mu/100) * glycogene_human_cell.df$samples
glycogene_human_cell.df$pseudopresence.pred_sd <- (predicted_percent_pseudopresence$s/100) * glycogene_human_cell.df$samples

```

#### computing pseudopresence pvalue

```{r}

glycogene_human_cell.df$pseudopresence.pval <-
  compute_pseudopresence_pvalue(
    glycogene_human_cell.df$pseudopresence,
    glycogene_human_cell.df$pseudopresence.pred_mu,
    glycogene_human_cell.df$pseudopresence.pred_sd,
    nrow(pseudopresence_model.lm$glycogene_sd_fit.lm$model))

```

#### computing expected pseudopresence

```{r}

glycogene_human_cell.df$pseudopresence.pred_mu <- with(
  glycogene_human_cell.df,
  round(pseudopresence.pred_mu, 2))

glycogene_human_cell.df$pseudopresence.pred_sd <- with(
  glycogene_human_cell.df,
  round(pseudopresence.pred_sd, 2))

glycogene_human_cell.df <- tidyr::unite(
  glycogene_human_cell.df,
  'pseudopresence.expected',
  pseudopresence.pred_mu,
  pseudopresence.pred_sd,
  sep = '+/-')

```

#### computing pseudopresence classification

```{r}

glycogene_human_cell.df$pseudopresence.classification <- with(
  glycogene_human_cell.df,
  ifelse(pseudopresence.pval > 1E-5, 'stable','unstable'))

```

#### computing intersect of single cell on single cell

```{r}

test_data <- glycogene_human_cell.df[
  glycogene_human_cell.df$gene %in% intersect(
    unique(glycogene_human_cell.df$gene),
    unique(glycogene_human_cell.df$gene)),]

reference_data <- test_data

glycogene_human_cell.df <- merge(
  glycogene_human_cell.df,
  rename(
    compute_intersects(
      test_data = reshape2:::dcast(
        test_data,
        gene ~ cell,
        value.var = 'rel_diff'),
      reference_data = reshape2:::dcast(
        reference_data,
        gene ~ cell,
        value.var = 'rel_diff'))$predicted_expression,
    cell=subset, sc_percentile=percentile),
  by = c('cell','gene'),
  all.x = T)

```

```{r}

glycogene_human_cell.df$sc_classification <- with(
  glycogene_human_cell.df, cut.default(
    sc_percentile,
    breaks = seq(-0.1,1.1,0.1),
    labels = c(paste0('<',seq(0,100,10),'%'),'>100%')))

```

#### computing intersect of single cell on bulk

```{r}

test_data <- glycogene_human_cell.df[
  glycogene_human_cell.df$gene %in% intersect(
    unique(glycogene_human_cell.df$gene),
    unique(bulk$gene[bulk$source == 'GTEx'])),]

reference_data <- bulk[
  bulk$source == 'GTEx' &
  bulk$gene %in% intersect(
    unique(bulk$gene[bulk$source == 'GTEx']),
    unique(glycogene_human_cell.df$gene)),]

glycogene_human_cell.df <- merge(
  glycogene_human_cell.df,
  rename(
    compute_intersects(
      test_data = reshape2:::dcast(
        test_data,
        gene ~ cell,
        value.var = 'rel_diff'),
      reference_data = reshape2:::dcast(
        bulk,
        gene ~ paste0(sample_id,':',tissue,':',source),
        value.var = 'rel_diff'))$predicted_expression,
    cell=subset, bulk_percentile=percentile),
  by = c('cell','gene'),
  all.x = T)

```

```{r}

glycogene_human_cell.df$bulk_classification <- with(
  glycogene_human_cell.df, cut.default(
    bulk_percentile,
    breaks = seq(-0.1,1.1,0.1),
    labels = c(paste0('<',seq(0,100,10),'%'),'>100%')))

```

### renaming columns

```{r}

glycogene_human.df <- dplyr::rename(
  glycogene_human.df,
  `gene ortholog` = gene_ortholog,
  `# of cells` = samples,
  `above cutoff` = above_cutoff,
  `above cutoff.pval` = pvalue,
  `above cutoff.classification` = `status5E-03`,
  `# of cells with transcript` = pseudopresence,
  `expected # of cells with transcript` = pseudopresence.expected,
  `# of cells with transcript.pval` = pseudopresence.pval,
  `# of cells with transcript.classification` = pseudopresence.classification,
  `seg normalised` = rel_diff,
  `sc %ile` = sc_percentile,
  `sc %ile.classification` = sc_classification,
  `bulk %ile` = bulk_percentile,
  `bulk %ile.classification` = bulk_classification)

```

```{r}

glycogene_human_cell.df <- dplyr::rename(
  glycogene_human_cell.df,
  `gene ortholog` = gene_ortholog,
  `# of cells` = samples,
  `above cutoff` = above_cutoff,
  `above cutoff.pval` = pvalue,
  `above cutoff.classification` = `status5E-03`,
  `# of cells with transcript` = pseudopresence,
  `expected # of cells with transcript` = pseudopresence.expected,
  `# of cells with transcript.pval` = pseudopresence.pval,
  `# of cells with transcript.classification` = pseudopresence.classification,
  `seg normalised` = rel_diff,
  `sc %ile` = sc_percentile,
  `sc %ile.classification` = sc_classification,
  `bulk %ile` = bulk_percentile,
  `bulk %ile.classification` = bulk_classification)

```

### ordering columns

```{r}

glycogene_human.df <- dplyr::select(
  glycogene_human.df,
  organism,
  tissue,
  cell,
  gene,
  `gene ortholog`,
  `# of cells`,
  pseudobulk,
  `above cutoff`,
  `above cutoff.pval`,
  `above cutoff.classification`,
  `# of cells with transcript`,
  `expected # of cells with transcript`,
  `# of cells with transcript.pval`,
  `# of cells with transcript.classification`,
  clr,
  `seg normalised`,
  `sc %ile`,
  `sc %ile.classification`,
  `bulk %ile`,
  `bulk %ile.classification`)

```

```{r}

glycogene_human_cell.df <- dplyr::select(
  glycogene_human_cell.df,
  organism,
  cell,
  gene,
  `gene ortholog`,
  `# of cells`,
  pseudobulk,
  `above cutoff`,
  `above cutoff.pval`,
  `above cutoff.classification`,
  `# of cells with transcript`,
  `expected # of cells with transcript`,
  `# of cells with transcript.pval`,
  `# of cells with transcript.classification`,
  clr,
  `seg normalised`,
  `sc %ile`,
  `sc %ile.classification`,
  `bulk %ile`,
  `bulk %ile.classification`)

```

# PanglaoDB (mouse)

## loading seg normalised data (glycogenes)

```{r}

file.copy('../../preprocess/generate_single_cell_data/pseudobulk_log_seg_norm.Rds', 'pseudobulk_log_seg_norm.Rds')
glycogene_segnorm.df <- readRDS('pseudobulk_log_seg_norm.Rds')
file.remove('pseudobulk_log_seg_norm.Rds')

```

```{r}

glycogene_mouse_segnorm.df <- glycogene_segnorm.df[with(glycogene_segnorm.df, source == 'mouse'),]

```

```{r}
glycogene_mouse_segnorm.df$tissue <- attributes(glycogene_segnorm.df)$cluster[glycogene_segnorm.df$source == 'mouse']
```

```{r}

remove(glycogene_segnorm.df)

```

## loading clrs (all genes)

```{r}

file.copy('../../preprocess/generate_single_cell_data/alldata_clean_mouse_clr_sparse.Rds', 'alldata_clean_mouse_clr_sparse.Rds')
alldata_clean_mouse_clr_sparse.sm <- readRDS('alldata_clean_mouse_clr_sparse.Rds')
file.remove('alldata_clean_mouse_clr_sparse.Rds')

```

### filtering for glycogenes only

```{r}

glycogene_clean_mouse_clr_sparse.sm = alldata_clean_mouse_clr_sparse.sm[,c(intersect(ortho_trans_table[glycogenes], colnames(alldata_clean_mouse_clr_sparse.sm)),'Dag1')]

```

```{r}

colnames(glycogene_clean_mouse_clr_sparse.sm) = dplyr::recode(colnames(glycogene_clean_mouse_clr_sparse.sm), Large='Large1', Gyltl1b='Large2')

```

```{r}

remove(alldata_clean_mouse_clr_sparse.sm)

```

### converting to data.frame

```{r}

glycogene_mouse_clr.df <-
  setNames(
    subset(reshape2:::melt.matrix(glycogene_clean_mouse_clr_sparse.sm, as.is = T), value != 0),
    c('tissue','gene','clr'))

```

```{r}

glycogene_mouse_clr.df$gene <- toupper(glycogene_mouse_clr.df$gene)

```

```{r}

remove(glycogene_clean_mouse_clr_sparse.sm)

```

## loading pseudobulks (glycogenes)

```{r}

file.copy('../../preprocess/generate_single_cell_data/glycogene_mouse_panglaodb_pseudobulkpresence.Rds', 'glycogene_mouse_panglaodb_pseudobulkpresence.Rds')
glycogene_mouse_pseudobulkpresence.df <- readRDS('glycogene_mouse_panglaodb_pseudobulkpresence.Rds')
file.remove('glycogene_mouse_panglaodb_pseudobulkpresence.Rds')

```

### relabeling clusters to match seg norm data

```{r}

glycogene_mouse_pseudobulkpresence.df <- dplyr::rename(glycogene_mouse_pseudobulkpresence.df, tissue=cluster, samples=cellnum)

```

```{r}

glycogene_mouse_pseudobulkpresence.df$gene <- toupper(glycogene_mouse_pseudobulkpresence.df$gene)

```

### merging datasets

```{r}

glycogene_mouse.df <- Reduce(
  merge,
  list(
    glycogene_mouse_pseudobulkpresence.df,
    glycogene_mouse_clr.df,
    glycogene_mouse_segnorm.df))

```

```{r}

remove(glycogene_mouse_pseudobulkpresence.df)
remove(glycogene_mouse_clr.df)
remove(glycogene_mouse_segnorm.df)

```

### averaging across tissue and cell type

<!-- dropping SRS, tsne, status5E-03, pseudopresence.pred_mu, pseudopresence.pred_sd  -->
<!-- and recomputing as averaged across tissue and cell because hiren can't make up his goddamn mind -->

```{r}

glycogene_mouse.df <- tidyr::separate(glycogene_mouse.df, 'tissue', into = c('SRS','tissue','cell'), sep = ':')
glycogene_mouse.df <- tidyr::separate(glycogene_mouse.df, 'cell', into = c('cell','tsne'), sep = '_cluster_')

glycogene_mouse.df$SRS <- NULL
glycogene_mouse.df$tsne <- NULL

glycogene_mouse.df <-
  group_by(glycogene_mouse.df, tissue, cell, gene) %>%
  summarise(
    samples = sum(samples),
    pseudobulk = mean(pseudobulk),
    clr = mean(clr),
    rel_diff = mean(rel_diff))

glycogene_mouse.df$pvalue <- with(
  glycogene_mouse.df,
  compute_expression_pvalue(pseudobulk, compute_threshold(samples)))

glycogene_mouse.df$`status5E-03` <- with(
  glycogene_mouse.df,
  addNA(factor(
    ifelse(
      pvalue > 5E-03, 'unpredictable',
      ifelse(pseudobulk > cutoff.int, 'expressed', 'unexpressed')),
    levels = rev(c('unexpressed','unpredictable','expressed')))))

glycogene_mouse.df$pseudopresence.pred_mu <- with(
  glycogene_mouse.df,
  (compute_sample_pop(
    pseudobulk,
    pseudopresence_model.lm$glycogene_mean_fit.lm,
    pseudopresence_model.lm$glycogene_sd_fit.lm)$mu/
    100)*samples)

glycogene_mouse.df$pseudopresence.pred_sd <- with(
  glycogene_mouse.df,
  (compute_sample_pop(
    pseudobulk,
    pseudopresence_model.lm$glycogene_mean_fit.lm,
    pseudopresence_model.lm$glycogene_sd_fit.lm)$s/
    100)*samples)

```

### joining columns

```{r}

glycogene_mouse.df$pseudopresence.pred_mu <- with(
  glycogene_mouse.df,
  round(pseudopresence.pred_mu, 2))

glycogene_mouse.df$pseudopresence.pred_sd <- with(
  glycogene_mouse.df,
  round(pseudopresence.pred_sd, 2))

glycogene_mouse.df <- tidyr::unite(
  glycogene_mouse.df,
  'pseudopresence.expected',
  pseudopresence.pred_mu,
  pseudopresence.pred_sd,
  sep = '+/-')

```

### adding columns

```{r}

glycogene_mouse.df$organism <- 'mouse'

```

```{r}

glycogene_mouse.df$gene_ortholog <- unname(ortho_trans_table[glycogene_mouse.df$gene])

```

```{r}

glycogene_mouse.df$above_cutoff <- glycogene_mouse.df$pseudobulk > cutoff.int

```

```{r}

glycogene_mouse_known.df <- subset(glycogene_mouse.df, cell != 'Unknown')

```

```{r}

test_data <- glycogene_mouse_known.df[
  glycogene_mouse_known.df$gene %in% intersect(
    unique(glycogene_mouse_known.df$gene),
    unique(glycogene_mouse_known.df$gene)),]

reference_data <- test_data

glycogene_mouse_known.df <- merge(
  glycogene_mouse_known.df,
  compute_intersects(
    test_data = reshape2:::dcast(
      test_data,
      gene ~ paste0(tissue,':',cell),
      value.var = 'rel_diff'),
    reference_data = reshape2:::dcast(
      reference_data,
      gene ~ paste0(tissue,':',cell),
      value.var = 'rel_diff'))$predicted_expression %>%
    tidyr::separate('subset', into = c('tissue', 'cell'), sep = ':') %>%
    dplyr::rename(sc_percentile=percentile),
  by = c('tissue','cell','gene'),
  all.x = T)

```

```{r}

glycogene_mouse_known.df$sc_classification <- with(
  glycogene_mouse_known.df, cut.default(
    sc_percentile,
    breaks = seq(-0.1,1.1,0.1),
    labels = c(paste0('<',seq(0,100,10),'%'),'>100%')))

```

```{r}

test_data <- glycogene_mouse_known.df[
  glycogene_mouse_known.df$gene %in% intersect(
    unique(glycogene_mouse_known.df$gene),
    unique(bulk$gene[bulk$source == 'GTEx'])),]

reference_data <- bulk[
  bulk$source == 'GTEx' &
  bulk$gene %in% intersect(
    unique(bulk$gene[bulk$source == 'GTEx']),
    unique(glycogene_mouse_known.df$gene)),]

glycogene_mouse_known.df <- merge(
  glycogene_mouse_known.df,
  compute_intersects(
    test_data = reshape2:::dcast(
      test_data,
      gene ~ paste0(tissue,':',cell),
      value.var = 'rel_diff'),
    reference_data = reshape2:::dcast(
      reference_data,
      gene ~ paste0(sample_id,':',tissue,':',source),
      value.var = 'rel_diff'))$predicted_expression %>%
    tidyr::separate('subset', into = c('tissue', 'cell'), sep = ':') %>%
    dplyr::rename(bulk_percentile=percentile),
  by = c('tissue','cell','gene'),
  all.x = T)

```

```{r}

glycogene_mouse_known.df$bulk_classification <- with(
  glycogene_mouse_known.df, cut.default(
    bulk_percentile,
    breaks = seq(-0.1,1.1,0.1),
    labels = c(paste0('<',seq(0,100,10),'%'),'>100%')))

```

```{r}

glycogene_mouse_unknown.df <- subset(glycogene_mouse.df, cell == 'Unknown')

```

```{r}

test_data <- glycogene_mouse_unknown.df[
  glycogene_mouse_unknown.df$gene %in% intersect(
    unique(glycogene_mouse_unknown.df$gene),
    unique(glycogene_mouse_unknown.df$gene)),]

reference_data <- test_data

glycogene_mouse_unknown.df <- merge(
  glycogene_mouse_unknown.df,
  compute_intersects(
    test_data = reshape2:::dcast(
      test_data,
      gene ~ paste0(tissue,':',cell),
      value.var = 'rel_diff'),
    reference_data = reshape2:::dcast(
      reference_data,
      gene ~ paste0(tissue,':',cell),
      value.var = 'rel_diff'))$predicted_expression %>%
    tidyr::separate('subset', into = c('tissue', 'cell'), sep = ':') %>%
    dplyr::rename(sc_percentile=percentile),
  by = c('tissue','cell','gene'),
  all.x = T)

```

```{r}

glycogene_mouse_unknown.df$sc_classification <- with(
  glycogene_mouse_unknown.df, cut.default(
    sc_percentile,
    breaks = seq(-0.1,1.1,0.1),
    labels = c(paste0('<',seq(0,100,10),'%'),'>100%')))

```

```{r}

test_data <- glycogene_mouse_unknown.df[
  glycogene_mouse_unknown.df$gene %in% intersect(
    unique(glycogene_mouse_unknown.df$gene),
    unique(bulk$gene[bulk$source == 'GTEx'])),]

reference_data <- bulk[
  bulk$source == 'GTEx' &
  bulk$gene %in% intersect(
    unique(bulk$gene[bulk$source == 'GTEx']),
    unique(glycogene_mouse_unknown.df$gene)),]

glycogene_mouse_unknown.df <- merge(
  glycogene_mouse_unknown.df,
  compute_intersects(
    test_data = reshape2:::dcast(
      test_data,
      gene ~ paste0(tissue,':',cell),
      value.var = 'rel_diff'),
    reference_data = reshape2:::dcast(
      reference_data,
      gene ~ paste0(sample_id,':',tissue,':',source),
      value.var = 'rel_diff'))$predicted_expression %>%
    tidyr::separate('subset', into = c('tissue', 'cell'), sep = ':') %>%
    dplyr::rename(bulk_percentile=percentile),
  by = c('tissue','cell','gene'),
  all.x = T)

```

```{r}

glycogene_mouse_unknown.df$bulk_classification <- with(
  glycogene_mouse_unknown.df, cut.default(
    bulk_percentile,
    breaks = seq(-0.1,1.1,0.1),
    labels = c(paste0('<',seq(0,100,10),'%'),'>100%')))

```

```{r}

glycogene_mouse.df <- rbind.data.frame(
  glycogene_mouse_known.df,
  glycogene_mouse_unknown.df)

```

```{r}

remove(glycogene_mouse_known.df)
remove(glycogene_mouse_unknown.df)

```

### repeating for cell specific aggregate

#### subset the dataset

```{r}

glycogene_mouse_cell.df <-
  glycogene_mouse.df %>% group_by(cell, gene) %>% summarize(
    organism = unique(organism),
    gene_ortholog = unique(gene_ortholog),
    clr = mean(clr),
    pseudobulk = mean(pseudobulk),
    samples = sum(samples),
    rel_diff = mean(rel_diff))

```

#### apply the cutoff

```{r}

glycogene_mouse_cell.df$above_cutoff <- glycogene_mouse_cell.df$pseudobulk > cutoff.int

```

#### compute the threshold

```{r}
glycogene_mouse_cell.df$threshold <- compute_threshold(glycogene_mouse_cell.df$samples)
```

#### compute the expression pvalue

```{r}
glycogene_mouse_cell.df$pvalue <- compute_expression_pvalue(glycogene_mouse_cell.df$pseudobulk, glycogene_mouse_cell.df$threshold)
glycogene_mouse_cell.df$threshold <- NULL
```

#### compute the expression classification

```{r}
glycogene_mouse_cell.df$`status5E-03` <- with(
  glycogene_mouse_cell.df,
  addNA(factor(
    ifelse(
      pvalue > 5E-03, 'unpredictable',
      ifelse(pseudobulk > cutoff.int, 'expressed', 'unexpressed')),
    levels = rev(c('unexpressed','unpredictable','expressed')))))
```

#### computing expected pseudopresence mu and sd

```{r}
predicted_percent_pseudopresence <-
  compute_sample_pop(
    glycogene_mouse_cell.df$pseudobulk,
    pseudopresence_model.lm$glycogene_mean_fit.lm,
    pseudopresence_model.lm$glycogene_sd_fit.lm)

```

```{r}

glycogene_mouse_cell.df$pseudopresence.pred_mu <- (predicted_percent_pseudopresence$mu/100) * glycogene_mouse_cell.df$samples
glycogene_mouse_cell.df$pseudopresence.pred_sd <- (predicted_percent_pseudopresence$s/100) * glycogene_mouse_cell.df$samples

```

#### computing expected pseudopresence

```{r}

glycogene_mouse_cell.df$pseudopresence.pred_mu <- with(
  glycogene_mouse_cell.df,
  round(pseudopresence.pred_mu, 2))

glycogene_mouse_cell.df$pseudopresence.pred_sd <- with(
  glycogene_mouse_cell.df,
  round(pseudopresence.pred_sd, 2))

glycogene_mouse_cell.df <- tidyr::unite(
  glycogene_mouse_cell.df,
  'pseudopresence.expected',
  pseudopresence.pred_mu,
  pseudopresence.pred_sd,
  sep = '+/-')

```

#### computing intersect of single cell on single cell

```{r}

test_data <- glycogene_mouse_cell.df[
  glycogene_mouse_cell.df$gene %in% intersect(
    unique(glycogene_mouse_cell.df$gene),
    unique(glycogene_mouse_cell.df$gene)),]

reference_data <- test_data

glycogene_mouse_cell.df <- merge(
  glycogene_mouse_cell.df,
  rename(
    compute_intersects(
      test_data = reshape2:::dcast(
        test_data,
        gene ~ cell,
        value.var = 'rel_diff'),
      reference_data = reshape2:::dcast(
        reference_data,
        gene ~ cell,
        value.var = 'rel_diff'))$predicted_expression,
    cell=subset, sc_percentile=percentile),
  by = c('cell','gene'),
  all.x = T)

```

```{r}

glycogene_mouse_cell.df$sc_classification <- with(
  glycogene_mouse_cell.df, cut.default(
    sc_percentile,
    breaks = seq(-0.1,1.1,0.1),
    labels = c(paste0('<',seq(0,100,10),'%'),'>100%')))

```

#### computing intersect of single cell on bulk

```{r}

test_data <- glycogene_mouse_cell.df[
  glycogene_mouse_cell.df$gene %in% intersect(
    unique(glycogene_mouse_cell.df$gene),
    unique(bulk$gene[bulk$source == 'GTEx'])),]

reference_data <- bulk[
  bulk$source == 'GTEx' &
  bulk$gene %in% intersect(
    unique(bulk$gene[bulk$source == 'GTEx']),
    unique(glycogene_mouse_cell.df$gene)),]

glycogene_mouse_cell.df <- merge(
  glycogene_mouse_cell.df,
  rename(
    compute_intersects(
      test_data = reshape2:::dcast(
        test_data,
        gene ~ cell,
        value.var = 'rel_diff'),
      reference_data = reshape2:::dcast(
        bulk,
        gene ~ paste0(sample_id,':',tissue,':',source),
        value.var = 'rel_diff'))$predicted_expression,
    cell=subset, bulk_percentile=percentile),
  by = c('cell','gene'),
  all.x = T)

```

```{r}

glycogene_mouse_cell.df$bulk_classification <- with(
  glycogene_mouse_cell.df, cut.default(
    bulk_percentile,
    breaks = seq(-0.1,1.1,0.1),
    labels = c(paste0('<',seq(0,100,10),'%'),'>100%')))

```

### renaming columns

```{r}

glycogene_mouse.df <- dplyr::rename(
  glycogene_mouse.df,
  `gene` = gene_ortholog,
  `gene ortholog` = gene,
  `# of cells` = samples,
  `above cutoff` = above_cutoff,
  `above cutoff.pval` = pvalue,
  `above cutoff.classification` = `status5E-03`,
  `expected # of cells with transcript` = pseudopresence.expected,
  `seg normalised` = rel_diff,
  `sc %ile` = sc_percentile,
  `sc %ile.classification` = sc_classification,
  `bulk %ile` = bulk_percentile,
  `bulk %ile.classification` = bulk_classification)

```

```{r}

glycogene_mouse_cell.df <- dplyr::rename(
  glycogene_mouse_cell.df,
  `gene` = gene_ortholog,
  `gene ortholog` = gene,
  `# of cells` = samples,
  `above cutoff` = above_cutoff,
  `above cutoff.pval` = pvalue,
  `above cutoff.classification` = `status5E-03`,
  `expected # of cells with transcript` = pseudopresence.expected,
  `seg normalised` = rel_diff,
  `sc %ile` = sc_percentile,
  `sc %ile.classification` = sc_classification,
  `bulk %ile` = bulk_percentile,
  `bulk %ile.classification` = bulk_classification)

```

### ordering columns

```{r}

glycogene_mouse.df <- dplyr::select(
  glycogene_mouse.df,
  organism,
  tissue,
  cell,
  gene,
  `gene ortholog`,
  `# of cells`,
  pseudobulk,
  `above cutoff`,
  `above cutoff.pval`,
  `above cutoff.classification`,
  `expected # of cells with transcript`,
  `clr`,
  `seg normalised`,
  `sc %ile`,
  `sc %ile.classification`,
  `bulk %ile`,
  `bulk %ile.classification`)

```

```{r}

glycogene_mouse_cell.df <- dplyr::select(
  glycogene_mouse_cell.df,
  organism,
  cell,
  gene,
  `gene ortholog`,
  `# of cells`,
  pseudobulk,
  `above cutoff`,
  `above cutoff.pval`,
  `above cutoff.classification`,
  `expected # of cells with transcript`,
  `clr`,
  `seg normalised`,
  `sc %ile`,
  `sc %ile.classification`,
  `bulk %ile`,
  `bulk %ile.classification`)

```

# computing dynamic ranges

### for GTEx samples

```{r}

GTEx_ranges <-
  compute_ranges(
    reference_data = reshape2:::dcast(
      GTEx_bulk,
      gene ~ paste0(sample_id,':',tissue),
      value.var = 'rel_diff'))

```

#### for TCGA samples

```{r}

TCGA_ranges <-
  compute_ranges(
    reference_data = reshape2:::dcast(
      TCGA_bulk,
      gene ~ paste0(sample_id,':',tissue),
      value.var = 'rel_diff'))

```

#### for all bulk samples

```{r}

bulk_ranges <-
  compute_ranges(
    reference_data = reshape2:::dcast(
      bulk,
      gene ~ paste0(sample_id,':',tissue,':',source),
      value.var = 'rel_diff'))

```

#### for Tabula Sapiens

```{r}

tabula_sapiens_ranges <-
  compute_ranges(
    reshape2:::dcast(
      glycogene_human.df,
      gene ~ paste0(tissue,':',cell),
      value.var = 'seg normalised'))

```

#### for Panglao Muris known

```{r}

panglao_musculus_known_ranges <-
  compute_ranges(
    reshape2:::dcast(
      subset(glycogene_mouse.df, cell != 'Unknown'),
      gene ~ paste0(tissue,':',cell),
      value.var = 'seg normalised'))
    # value.var = 'rel_diff))

```

#### for Panglao Muris unknown

```{r}

panglao_musculus_unknown_ranges <-
  compute_ranges(
    reshape2:::dcast(
      subset(glycogene_mouse.df, cell == 'Unknown'),
      gene ~ paste0(tissue,':',cell),
      value.var = 'seg normalised'))
    # value.var = 'rel_diff))

```

#### for Panglao Muris

```{r}

panglao_musculus_ranges <-
  compute_ranges(
    reshape2:::dcast(
      glycogene_mouse.df,
      gene ~ paste0(tissue,':',cell),
      value.var = 'seg normalised'))
    # value.var = 'rel_diff))

```

```{r}

glycopacity_internal_data <-
  list(
    bulk_ranges = bulk_ranges,
    TCGA_ranges = TCGA_ranges,
    GTEx_ranges = GTEx_ranges,
    tabula_sapiens_ranges = tabula_sapiens_ranges,
    panglao_musculus_unknown_ranges = panglao_musculus_unknown_ranges,
    panglao_musculus_known_ranges = panglao_musculus_known_ranges,
    panglao_musculus_ranges = panglao_musculus_ranges)

```

### writing rds

```{r}

attributes(glycopacity_internal_data)$repo = "glyco_atlas_data/generate_pseudobulkpresenceclrseg"
attributes(glycopacity_internal_data)$revision = system("echo \"`git symbolic-ref HEAD 2> /dev/null | cut -b 12-`-`git describe --always --dirty`\"",intern=T)
saveRDS(glycopacity_internal_data, 'glycopacity_internal_data.Rds')

```

```{r}

attributes(glycogene_human.df)$repo = "glyco_atlas_data/generate_pseudobulkpresenceclrseg"
attributes(glycogene_human.df)$revision = system("echo \"`git symbolic-ref HEAD 2> /dev/null | cut -b 12-`-`git describe --always --dirty`\"",intern=T)
saveRDS(glycogene_human.df, 'glycogene_human_pseudobulkpresence_clr_seg.Rds')

```

```{r}

attributes(glycogene_human_cell.df)$repo = "glyco_atlas_data/generate_pseudobulkpresenceclrseg"
attributes(glycogene_human_cell.df)$revision = system("echo \"`git symbolic-ref HEAD 2> /dev/null | cut -b 12-`-`git describe --always --dirty`\"",intern=T)
saveRDS(glycogene_human_cell.df, 'glycogene_human_pseudobulkpresence_clr_seg_cell.Rds')

```

```{r}

attributes(glycogene_mouse.df)$repo = "glyco_atlas_data/generate_pseudobulkpresenceclrseg"
attributes(glycogene_mouse.df)$revision = system("echo \"`git symbolic-ref HEAD 2> /dev/null | cut -b 12-`-`git describe --always --dirty`\"",intern=T)
saveRDS(glycogene_mouse.df, 'glycogene_mouse_pseudobulkpresence_clr_seg.Rds')

```

```{r}

attributes(glycogene_mouse_cell.df)$repo = "glyco_atlas_data/generate_pseudobulkpresenceclrseg"
attributes(glycogene_mouse_cell.df)$revision = system("echo \"`git symbolic-ref HEAD 2> /dev/null | cut -b 12-`-`git describe --always --dirty`\"",intern=T)
saveRDS(glycogene_mouse_cell.df, 'glycogene_mouse_pseudobulkpresence_clr_seg_cell.Rds')

```

### writing xlsx

```{r}

glycogene.df <- plyr::rbind.fill(glycogene_human.df, glycogene_mouse.df)
glycogene_cell.df <- plyr::rbind.fill(glycogene_human_cell.df, glycogene_mouse_cell.df)

```

```{r}

writexl::write_xlsx(
  x = list(
    by_tissue_cell = glycogene_human.df,
    by_cell = glycogene_human_cell.df),
  path = 'glycogene_human_pseudobulkpresence_clr_seg.xlsx')

```

```{r}

writexl::write_xlsx(
  x = list(
    by_tissue_cell = glycogene_mouse.df,
    by_cell = glycogene_mouse_cell.df),
  path = 'glycogene_mouse_pseudobulkpresence_clr_seg.xlsx')

```

```{r}

writexl::write_xlsx(
  x = list(
    by_tissue_cell = glycogene.df,
    by_cell = glycogene_cell.df),
  path = 'glycogene_pseudobulkpresence_clr_seg.xlsx')

```

